File: frontend/src/app/_services/training.service.ts
import {Injectable} from '@angular/core';
import {TrainingModel} from "../../models/training.model";
import {HttpClient, HttpHeaders} from "@angular/common/http";
import {Observable} from "rxjs";

@Injectable({
  providedIn: 'root'
})
export class TrainingService {
  // private host : string = "http://57.128.221.44:8888/TRAINING-SERVICE"
  private host : string = "http://51.254.114.223:8888/TRAINING-SERVICE"

  constructor(private http : HttpClient) { }

  public addTraining(training : TrainingModel) {
    return this.http.post<TrainingModel>(`${this.host}/training/add`, training)
  }

  public getTrainings() {
    return this.http.get<Array<TrainingModel>>(`${this.host}/training/all`)
  }

  public getTrainingsClient() {
    return this.http.get<Array<TrainingModel>>(`${this.host}/training/find/trainings-client`)
  }

  public getTrainingsByClient(idClient : number) {
    return this.http.get<Array<TrainingModel>>(`${this.host}/training/find/trainingsByClient/${idClient}`)
  }

  public getTraining(idTraining : number) {
    return this.http.get<TrainingModel>(`${this.host}/training/find/${idTraining}`)
  }

  public updateTraining(training : TrainingModel, idTraining : number) {
    return this.http.put<TrainingModel>(`${this.host}/training/update/${idTraining}`, training)
  }

  public updateLifeCycle(idTraining : number, training : TrainingModel) {
    return this.http.put<TrainingModel>(`${this.host}/training/updateLifeCycle/${idTraining}`, training)
  }

  public addPv(pv : string, idTraining : number) {
    return this.http.put<TrainingModel>(`${this.host}/training/add/pv/${idTraining}`, pv)
  }

  public removePv(idTraining : number, training : TrainingModel) {
    return this.http.put<TrainingModel>(`${this.host}/training/remove/pv/${idTraining}`, training)
  }

  public addTrainingSupport(formData: FormData, idTraining : number) {
    const headers = new HttpHeaders()
    headers.append('Content-Type', 'multipart/form-data')
    return this.http.put<TrainingModel>(`${this.host}/training/add/trainingSupport/${idTraining}`, formData, { headers })
  }

  public removeTrainingSupport(idTraining: number, training: TrainingModel) {
    return this.http.put<TrainingModel>(`${this.host}/training/remove/trainingSupport/${idTraining}`, training)
  }

  public addReferenceCertificate(formData: FormData, idTraining : number) {
    const headers = new HttpHeaders()
    headers.append('Content-Type', 'multipart/form-data')
    return this.http.put<TrainingModel>(`${this.host}/training/add/referenceCertificate/${idTraining}`, formData, { headers })
  }

  public removeReferenceCertificate(idTraining: number, training: TrainingModel) {
    return this.http.put<TrainingModel>(`${this.host}/training/remove/referenceCertificate/${idTraining}`, training)
  }

  public addTrainingNotes(trainingNotes: FormData, idTraining : number) {
    const headers = new HttpHeaders()
    headers.append('Content-Type', 'multipart/form-data')
    return this.http.put<TrainingModel>(`${this.host}/training/add/trainingNotes/${idTraining}`, trainingNotes, { headers })
  }

  public removeTrainingNotes(idTraining: number, training: TrainingModel) {
    return this.http.put<TrainingModel>(`${this.host}/training/remove/trainingNotes/${idTraining}`, training)
  }

  public deleteTraining(idTraining : number) {
    return this.http.delete<TrainingModel>(`${this.host}/training/delete/${idTraining}`)
  }

  public sendMail(emailData : any){
    return this.http.post<string>(`${this.host}/mail/send`, emailData, {responseType : 'text' as 'json'})
  }

  public checkIfPvExists(idTraining: number) : Observable<boolean> {
    return this.http.get<boolean>(`${this.host}/training/checkPv/${idTraining}`)
  }

  public checkIfTrainingSupportExists(idTraining: number) : Observable<boolean> {
    return this.http.get<boolean>(`${this.host}/training/checkTrainingSupport/${idTraining}`)
  }
}

File: training-service/src/main/java/org/example/trainingservice/service/TrainingService.java
package org.example.trainingservice.service;

import org.example.trainingservice.dto.AddTrainingDTO;
import org.example.trainingservice.dto.TrainingDTO;
import org.example.trainingservice.dto.TrainingInvoiceDTO;
import org.example.trainingservice.entity.Training;
import org.springframework.web.multipart.MultipartFile;

import java.util.List;

public interface TrainingService {
    AddTrainingDTO addTraining(AddTrainingDTO addTrainingDTO);

    List<TrainingDTO> getTrainings();

    TrainingDTO getById(Long idTraining);

    AddTrainingDTO updateTraining(AddTrainingDTO addTrainingDTO, Long idTraining);

    TrainingDTO updateStatus(Long idTraining, TrainingDTO trainingDTO, String status);

    AddTrainingDTO addPv(String pv, Long idTraining);

    TrainingDTO removePv(Long idTraining, TrainingDTO trainingDTO);

    AddTrainingDTO addTrainingSupport(MultipartFile trainingSupport, Long idTraining);

    TrainingDTO removeTrainingSupport(Long idTraining, TrainingDTO trainingDTO);

    AddTrainingDTO addReferenceCertificate(MultipartFile referenceCertificate, Long idTraining);

    TrainingDTO removeReferenceCertificate(Long idTraining, TrainingDTO trainingDTO);

    AddTrainingDTO addTrainingNotes(MultipartFile presenceList, MultipartFile evaluation, Long idTraining);

    TrainingDTO removeTrainingNotes(Long idTraining, TrainingDTO trainingDTO);

    TrainingDTO deleteTraining(Long idTraining);

    List<TrainingDTO> getTrainingsOfWeek();

    List<TrainingInvoiceDTO> getTrainingsClient();

    List<TrainingInvoiceDTO> getTrainingsByClient(Long idClient);

    TrainingDTO getGroupsByTraining(Long idTraining);

    boolean checkIfPvExists(Long idTraining);

    boolean checkIfTrainingSupportExists(Long idTraining);

    List<TrainingDTO> getTrainingsByGroupIds(List<Long> groupIds);
}
